onnxscript/function_libs/torch_lib/ops/core.py

```diff
@@ -3026,10 +3028,54 @@ def aten_imag(self: TensorType) -> TensorType:
     raise NotImplementedError()
 
 
-def aten_index(self: TensorType, indices: Optional[Sequence[TensorType]]) -> TensorType:
+def are_consecutive(lst):
+    if len(lst) == 0:
+        return True
+    else:
+        return sorted(lst) == list(range(min(lst), max(lst) + 1))
+
+
+def _is_none_in_middle(indices):
+    not_none_indices = [i for i, idx in enumerate(indices) if idx is not None]
+    return not are_consecutive(not_none_indices)
+
+
+@torch_op("aten::index", trace_only=True)
+def aten_index(self: TensorType, indices: Sequence[Optional[INT64]]) -> TensorType:
     """index.Tensor(Tensor self, Tensor?[] indices) -> Tensor"""
 
-    raise NotImplementedError()
+    ordered_indices = sorted(range(len(indices)), key=lambda i: (indices[i] is None, i))
+    ordered_indices += list(range(len(ordered_indices), len(self.shape)))
+
+    self = op.Transpose(self, perm=ordered_indices)
+
+    # Need broadcast concat lol.
+    not_none_indices = [idx for idx in indices if idx is not None]
+    broadcast_shape = np.broadcast_shapes(*[idx.shape for idx in not_none_indices])
```

Comment:
We may not have shapes here: need to assert or use alternatives like implicit broadcasting?

Reply:
yea that's kinda rough. If we don't, or if there are symbols, we need to dynamically broadcast, probably via `add`. However at some point later we need the rank of broadcasted tensor, which is needed to construct the `perm` for transpose. If we cannot get it statically, then the transpose becomes a headache.

Reply:
I think we can assume the rank. This is more relaxed than assuming shape